app/services/theme_settings_migrations_runner.rb

```diff
@@ -0,0 +1,174 @@
+# frozen_string_literal: true
+
+class ThemeSettingsMigrationsRunner
+  Migration =
+    Struct.new(:version, :name, :code, :theme_field_id) do
+      def original_name
+        "#{version.to_s.rjust(4, "0")}-#{name}"
+      end
+    end
+
+  MIGRATION_ENTRY_POINT_JS = <<~JS
+    const migrate = require("discourse/theme/migration")?.default;
+    function main(settingsObj) {
+      if (!migrate) {
+        throw new Error("no_exported_migration_function");
+      }
+      if (typeof migrate !== "function") {
+        throw new Error("default_export_is_not_a_function");
+      }
+      const map = new Map(Object.entries(settingsObj));
+      const updatedMap = migrate(map);
+      if (!updatedMap) {
+        throw new Error("migration_function_no_returned_value");
+      }
+      if (!(updatedMap instanceof Map)) {
+        throw new Error("migration_function_wrong_return_type");
+      }
+      return Object.fromEntries(updatedMap.entries());
+    }
+  JS
+
+  private_constant :Migration, :MIGRATION_ENTRY_POINT_JS
+
+  def initialize(theme, limit: 100, timeout: 100, memory: 100.megabytes)
```

Comment:
Just my gut feel but It feels like 100 mb is way too much and might open up new security vectors for us. Is there a reason we chose 100mb to start with?

Reply:
Not really, I just picked a number and went with it. Maybe 20 MB is more sensible?

Reply:
For a simple migration, can we measure how much memory it uses? I think we should go as low as possible and then increase upwards if required. Slightly harder to start high and then reduce it.

Reply:
So Mini Racer contexts have a `heap_stats` method which spits out a few memory metrics, see https://github.com/rubyjs/mini_racer#controlling-memory. From some testing that I've done locally, a context starts out with around 800kb for the `used_heap_size` metric, then after I load up the loader.js library and the migration (the setting rename example in the OP) into the context and execute the migration, the metric grows by around 30kb. All other metrics remain unchanged.

Given the numbers I got, I think 2mb should be enough for now?

```ruby
# frozen_string_literal: true
# execute with the `rails runner script_path.rb` command

migration = <<~JS
  export default function migrate(settings) {
    settings.set("new_key", settings.get("old_key"));
    return settings;
  }
JS

c = MiniRacer::Context.new

t1 = c.heap_stats.dup

c.eval(ThemeSettingsMigrationsRunner.loader_js_lib_content, filename: "loader.js")
c.eval(
  DiscourseJsProcessor.transpile(migration, "", "discourse/theme/migration"),
  filename: "theme-233-migration.js",
)
c.eval(ThemeSettingsMigrationsRunner.const_get(:MIGRATION_ENTRY_POINT_JS), filename: "migration-entrypoint.js")
c.call("main", { "old_key" => "osama" })

t2 = c.heap_stats.dup

helper = ActionController::Base.helpers
t2.keys.each do |key|
  diff = t2[key] - t1[key]
  puts "#{key} was: #{helper.number_to_human_size(t1[key])} and grew by: #{helper.number_to_human_size(diff)}."
end
```

Reply:
2mb sounds good :+1: